<table style="width:100%">
<tr>
<td style="vertical-align:middle; text-align:left;">
<font size="2">
Supplementary code for the <a href="http://mng.bz/orYv">Build a Large Language Model From Scratch</a> book by <a href="https://sebastianraschka.com">Sebastian Raschka</a><br>
<br>Code repository: <a href="https://github.com/rasbt/LLMs-from-scratch">https://github.com/rasbt/LLMs-from-scratch</a>
<br>汉化的库: <a href="https://github.com/GoatCsu/CN-LLMs-from-scratch.git">https://github.com/GoatCsu/CN-LLMs-from-scratch.git</a>
</font>
</td>
<td style="vertical-align:middle; text-align:left;">
<a href="http://mng.bz/orYv"><img src="https://sebastianraschka.com/images/LLMs-from-scratch-images/cover-small.webp" width="100px"></a>
</td>
</tr>
</table>


# 附录A: Pytorch介绍

许多科学计算库不会立即支持最新版本的 Python。因此，在安装 PyTorch 时，建议使用
比最新版本旧一到两个版本的 Python。如果最新的 Python 版本是 Python 3.13，那么推荐使用
Python 3.11 或 Python 3.12

`pip install torch`

假设你的计算机支持兼容 CUDA 的 GPU。在这种情况下，如果你正在使用的 Python 环境已安装
必要的依赖项（如 pip），那么系统将自动安装支持 CUDA 加速的 PyTorch 版本。


本书中使用的是 PyTorch 2.4.0，为了确保与本书的兼容性，建议你使用以下命令安装该版本：
`pip install torch==2.4.0`

建议你访问 PyTorch 官方网站并使用安装菜单选择适合你操作系统的安装命令

In [ ]:
import sys
print(sys.version)

3.12.11 (main, Jun  4 2025, 08:56:18) [GCC 11.4.0]


## A.1 什么是PyTorch

In [ ]:
import torch

print(torch.__version__)

2.8.0+cu126


In [ ]:
print(torch.cuda.is_available())

False


<img src="https://raw.githubusercontent.com/MLNLP-World/LLMs-from-scratch-CN/main/imgs/A1/1.png" width="400px">

<img src="https://raw.githubusercontent.com/MLNLP-World/LLMs-from-scratch-CN/main/imgs/A1/2.png" width="300px">

<img src="https://raw.githubusercontent.com/MLNLP-World/LLMs-from-scratch-CN/main/imgs/A1/3.png" width="300px">

<img src="https://raw.githubusercontent.com/MLNLP-World/LLMs-from-scratch-CN/main/imgs/A1/4.png" width="500px">

<img src="https://raw.githubusercontent.com/MLNLP-World/LLMs-from-scratch-CN/main/imgs/A1/5.png" width="500px">

## A.2 理解张量

<img src="https://raw.githubusercontent.com/MLNLP-World/LLMs-from-scratch-CN/main/imgs/A2/1.png" width="400px">

PyTorch 采用了大部分 NumPy 数组 API 和语法来进行张量操作, 如果你对 NumPy 不熟悉，可以通过我的文章:
Scientific Computing in Python: Introduction to NumPy and Matplotlib >>> https://sebastianraschka.com/blog/2020/numpy-intro.html

### A.2.1 标量、向量、矩阵和张量

In [ ]:
import torch
import numpy as np

# 从Python整数创建一个零维张量（标量）
tensor0d = torch.tensor(1)

# 从Python列表创建一个一维张量（向量）
tensor1d = torch.tensor([1, 2, 3])

# 从嵌套的Python列表创建一个二维张量
tensor2d = torch.tensor([[1, 2],
                         [3, 4]])

# 从嵌套的Python列表创建一个三维张量
tensor3d_1 = torch.tensor([[[1, 2], [3, 4]],
                           [[5, 6], [7, 8]]])

# 从NumPy数组创建一个张量
ary3d = np.array([[[1, 2], [3, 4]],
                  [[5, 6], [7, 8]]])
tensor3d_2 = torch.tensor(ary3d)  # 复制 NumPy 数组
tensor3d_3 = torch.from_numpy(ary3d)  # 与 NumPy 数组共享内存

In [ ]:
print(tensor0d)
print(tensor1d)
print(tensor2d)
print(tensor3d_1)
print(tensor3d_2)
print(tensor3d_3)

tensor(1)
tensor([1, 2, 3])
tensor([[1, 2],
        [3, 4]])
tensor([[[1, 2],
         [3, 4]],

        [[5, 6],
         [7, 8]]])
tensor([[[1, 2],
         [3, 4]],

        [[5, 6],
         [7, 8]]])
tensor([[[1, 2],
         [3, 4]],

        [[5, 6],
         [7, 8]]])


In [ ]:
ary3d[0, 0, 0] = 999
print(tensor3d_2) # 保持不变

tensor([[[1, 2],
         [3, 4]],

        [[5, 6],
         [7, 8]]])


In [ ]:
print(tensor3d_3) # 由于共享内存而发生变化

tensor([[[999,   2],
         [  3,   4]],

        [[  5,   6],
         [  7,   8]]])


### A.2.2 张量数据类型

In [ ]:
tensor1d = torch.tensor([1, 2, 3])
print(tensor1d.dtype) #查看数据类型 PyTorch 采用 Python 默认的 64 位整数数据类型

torch.int64


In [ ]:
floatvec = torch.tensor([1.0, 2.0, 3.0])
print(floatvec.dtype) #如果使用 Python 浮点数创建张量，那么 PyTorch 默认会创建具有 32 位精度的张量, 这种选择主要是为了在精度和计算效率之间取得平衡, 1. 大多数深度学习任务里足够精度, 且内存和计算资源少 2. GPU架构对32bit有优化

torch.float32


In [ ]:
floatvec = tensor1d.to(torch.float32) # 可以使用张量的.to 方法更改精度
print(floatvec.dtype)

torch.float32


### A.2.3 常见的PyTorch张量操作

In [ ]:
tensor2d = torch.tensor([[1, 2, 3],
                         [4, 5, 6]])
tensor2d

tensor([[1, 2, 3],
        [4, 5, 6]])

In [ ]:
tensor2d.shape # .shape 属性允许我们访问张量的形状

torch.Size([2, 3])

In [ ]:
tensor2d.reshape(3, 2) #要将该张量变为 3×2 的形状

tensor([[1, 2],
        [3, 4],
        [5, 6]])

In [ ]:
tensor2d.view(3, 2) #在 PyTorch 中，重塑张量更常用的命令是.view()

tensor([[1, 2],
        [3, 4],
        [5, 6]])

In [ ]:
print(tensor2d)

tensor([[1, 2, 3],
        [4, 5, 6]])


.view()和.reshape()的微妙区别在于它们对内存布局的处理方式：.view()要求原始数据是连续的，如果不是，它将无法工作，而.reshape()会工作，如有必要，它会复制数据以确保所需的形状

In [ ]:
tensor2d.T #转置张量, 将其沿对角线翻转

tensor([[1, 4],
        [2, 5],
        [3, 6]])

In [ ]:
tensor2d.matmul(tensor2d.T) #PyTorch 中常用的矩阵相乘方法, 也可以使用@运算符，它能够更简洁地实现相同的功能

tensor([[14, 32],
        [32, 77]])

In [ ]:
tensor2d @ tensor2d.T

tensor([[14, 32],
        [32, 77]])

## A.3 将模型视为计算图

<img src="https://raw.githubusercontent.com/MLNLP-World/LLMs-from-scratch-CN/main/imgs/A3/1.png" width="600px">

In [ ]:
import torch.nn.functional as F

y = torch.tensor([1.0])  # 真实标签
x1 = torch.tensor([1.1]) # 输入特征
w1 = torch.tensor([2.2]) # 权重参数
b = torch.tensor([0.0])  # 偏置单元

z = x1 * w1 + b          # 净输入
a = torch.sigmoid(z)     # 激活和输出

loss = F.binary_cross_entropy(a, y)
print(loss)

tensor(0.0852)


## A.4 轻松实现自动微分

<img src="https://raw.githubusercontent.com/MLNLP-World/LLMs-from-scratch-CN/main/imgs/A4/1.png" width="600px">

In [ ]:
import torch.nn.functional as F #Imports the functional interface for neural network operations from PyTorch.
from torch.autograd import grad #Imports the grad function specifically for calculating gradients.

y = torch.tensor([1.0]) #Defines the true label as a PyTorch tensor with a value of 1.0.
x1 = torch.tensor([1.1]) #Defines the input feature as a PyTorch tensor with a value of 1.1.
w1 = torch.tensor([2.2], requires_grad=True) #Defines the weight parameter as a PyTorch tensor with a value of 2.2. requires_grad=True is crucial here, as it tells PyTorch to track operations on this tensor so that gradients can be computed later.
b = torch.tensor([0.0], requires_grad=True) #Defines the bias term as a PyTorch tensor with a value of 0.0

z = x1 * w1 + b  #Calculates the net input z by performing a linear transformation on the input feature x1 using the weight w1 and adding the bias b.
a = torch.sigmoid(z) #Applies the sigmoid activation function to the net input z to get the output a

loss = F.binary_cross_entropy(a, y) #Calculates the binary cross-entropy loss between the output a and the true label y

grad_L_w1 = grad(loss, w1, retain_graph=True) #默认情况下，PyTorch 在计算梯度后会销毁计算图以释放内存。然而，由于我们即将再次使用这个计算图，因此可以设置 retain_graph=True，使其保留在内存中
grad_L_b = grad(loss, b, retain_graph=True)

print(grad_L_w1)
print(grad_L_b)
#这里我们手动使用了 grad 函数，这在实验、调试和概念演示中很有用。

(tensor([-0.0898]),)
(tensor([-0.0817]),)


In [ ]:
# 在实际操作中，PyTorch 提供了更高级的工具来自动化这个过程。例如，我们可以对损失函数调用.backward方法，随后 PyTorch 将计算计算图中所有叶节点的梯度，这些梯度将通过张量的.grad 属性进行存储
loss.backward()

print(w1.grad)
print(b.grad)

#PyTorch 通过.backward方法为我们处理了微积分问题——我们不需要手动计算任何导数或梯度

tensor([-0.0898])
tensor([-0.0817])


## A.5 实现多层神经网络

<img src="https://raw.githubusercontent.com/MLNLP-World/LLMs-from-scratch-CN/main/imgs/A5/1.png" width="500px">

In [ ]:
'''
In summary, this class creates a simple feedforward neural network with two hidden layers
using ReLU activation and an output layer. The number of input features and output units
are customizable through the constructor.
'''
class NeuralNetwork(torch.nn.Module): #在 PyTorch 中实现神经网络时，可以通过子类化 torch.nn.Module 类来定义我们自己的自定义网络架构
    def __init__(self, num_inputs, num_outputs): # 将输入和输出的数量编码为变量，使我们可以在具有不同特征数量和类别数量的数据集上重复使用相同的代码
        super().__init__() #This line calls the constructor of the parent class (torch.nn.Module). This is necessary to properly initialize the module.

        self.layers = torch.nn.Sequential(

            # 1st hidden layer
            torch.nn.Linear(num_inputs, 30), #线性层将输入节点和输出节点的数量作为参数
            torch.nn.ReLU(), #非线性激活函数被放置在隐藏层之间

            # 2nd hidden layer
            torch.nn.Linear(30, 20), #一个隐藏层的输出节点数量必须与下一层的输入节点数量相匹配
            torch.nn.ReLU(),

            # output layer
            torch.nn.Linear(20, num_outputs),
        )

    def forward(self, x): #This method defines the forward pass of the neural network. It takes an input tensor x and passes it through the defined layers.
        logits = self.layers(x)
        return logits #最后一层的输出称为 logits

In [ ]:
model = NeuralNetwork(50, 3) #实例化一个新的神经网络对象
#(50, 3) are the arguments passed to the __init__ method of the NeuralNetwork class.

In [ ]:
'''
请注意，在实现 NeuralNetwork 类时，我们使用了 Sequential 类。虽然 Sequential 并非
必需，但如果有一系列想要按特定顺序执行的层（正如本例中的情况），那么使用它可以让我们
的工作更轻松。因此，在__init__构造函数中实例化 self.layers = Sequential(...)后，
只需在 NeuralNetwork 的 forward 方法中调用 self.layers，而无须单独调用每个层。
'''
print(model)

NeuralNetwork(
  (layers): Sequential(
    (0): Linear(in_features=50, out_features=30, bias=True)
    (1): ReLU()
    (2): Linear(in_features=30, out_features=20, bias=True)
    (3): ReLU()
    (4): Linear(in_features=20, out_features=3, bias=True)
  )
)


In [ ]:
#检查一下该模型的可训练参数总数
num_params = sum(p.numel() for p in model.parameters() if p.requires_grad) #每一个 requires_grad=True 的参数都会被视为可训练参数，并在训练期间进行更新
print("Total number of trainable model parameters:", num_params)

Total number of trainable model parameters: 2213


`num_params = sum(p.numel() for p in model.parameters() if p.requires_grad)`: This line calculates the sum of the number of elements (parameters) in all the layers of the `model` that require gradients.

- `model.parameters()`: This method returns an iterator over all the parameters (weights and biases) of the model.
- `for p in model.parameters()`: This iterates through each parameter p in the model.
if p.requires_grad: This condition filters the parameters to include only those that have `requires_grad` set to `True`. These are the parameters that will be updated during the training process.
- `p.numel()`: This method returns the total number of elements (scalars) in a given parameter tensor p.
- `sum(...)`: This sums up the number of elements for all the parameters that meet the condition.

对于前面我们提到的具有两个隐藏层的神经网络模型，这些可训练参数包含在 `torch.nn.Linear` 层中。`Linear` 层会将输入与权重矩阵相乘，并加上一个偏置向量。这有时被称为**前馈层**或**全连接层**

基于这里执行的 `print(model)`调用，可以看到第一个 `Linear` 层在 `layers` 属性中的索引位置是 0。

可以通过以下方式访问对应的权重参数矩阵

In [ ]:
print(model.layers[0].weight) #访问权重参数矩阵

Parameter containing:
tensor([[ 0.1388,  0.0159,  0.1215,  ...,  0.1032,  0.0296,  0.0102],
        [ 0.0229,  0.0260, -0.0458,  ..., -0.0358,  0.0362,  0.0497],
        [-0.0896,  0.0113,  0.1370,  ...,  0.1037,  0.1230, -0.0929],
        ...,
        [-0.1362, -0.0713, -0.0010,  ...,  0.1176,  0.1054, -0.1012],
        [ 0.1226,  0.0937, -0.1409,  ...,  0.1321, -0.0613,  0.0086],
        [-0.0045, -0.0604,  0.0535,  ...,  0.0697,  0.0373,  0.0923]],
       requires_grad=True)


In [ ]:
print(model.layers[0].weight.shape) #查看其维度
#这里的权重矩阵是一个 30×50 的矩阵，可以看到 requires_grad 被设置为 True（意味着该矩阵是可训练的）——这是 torch.nn.Linear 中权重和偏置的默认设置

torch.Size([30, 50])


In [ ]:
print(model.layers[0].bias) # 访问偏置向量

Parameter containing:
tensor([-0.0181,  0.1404,  0.0374,  0.1102,  0.0045,  0.0788,  0.1013,  0.0211,
         0.1191, -0.1204, -0.0152, -0.0222, -0.0056,  0.0466, -0.0365, -0.0321,
         0.0927, -0.1029, -0.0093,  0.1047,  0.1279, -0.1176,  0.0445,  0.0583,
         0.0263,  0.0459, -0.0549,  0.0258,  0.0305,  0.0463],
       requires_grad=True)


In [ ]:
print(model.layers[0].bias.shape)

torch.Size([30])


如果你在自己的计算机上执行前面的代码，那么权重矩阵中的数值可能会与本书展示的有所不同。

模型权重会用小的随机数进行初始化，每次实例化网络时这些数值都会不同。在深度学习中，使用小的随机数初始化模型权重是为了在训练过程中打破对称性。否则，各节点将执行相同的操作并在反向传播过程中进行相同的更新，导致网络无法学习从输入到输出的复杂映射关系。

In [ ]:
torch.manual_seed(123) #可以通过 manual_seed 来为 PyTorch 的随机数生成器设定种子，从而使随机数初始化可重复

model = NeuralNetwork(50, 3)
print(model.layers[0].weight)

Parameter containing:
tensor([[-0.0577,  0.0047, -0.0702,  ...,  0.0222,  0.1260,  0.0865],
        [ 0.0502,  0.0307,  0.0333,  ...,  0.0951,  0.1134, -0.0297],
        [ 0.1077, -0.1108,  0.0122,  ...,  0.0108, -0.1049, -0.1063],
        ...,
        [-0.0787,  0.1259,  0.0803,  ...,  0.1218,  0.1303, -0.1351],
        [ 0.1359,  0.0175, -0.0673,  ...,  0.0674,  0.0676,  0.1058],
        [ 0.0790,  0.1343, -0.0293,  ...,  0.0344, -0.0971, -0.0509]],
       requires_grad=True)


In [ ]:
'''
This code snippet demonstrates how to perform a forward pass through the model
with a single input sample and how to prevent gradient calculation during inference.
'''

torch.manual_seed(123)

X = torch.rand((1, 50)) #This line creates a random tensor named X with a shape of (1, 50).
# torch.rand() creates a tensor filled with random numbers from a uniform distribution between 0 and 1.
# (1, 50) specifies the shape of the tensor. A shape of (1, 50) represents a single input sample with 50 features, which matches the num_inputs specified when the NeuralNetwork was instantiated earlier.

out = model(X) #This line performs a forward pass through the model using the input tensor X
#When you call a torch.nn.Module instance like a function (model(X)), it automatically executes the forward method you defined in the class. The output of the forward pass (the logits in this case) is stored in the out variable.
print(out)

#You'll notice that the output tensor has grad_fn=<AddmmBackward0>, indicating that PyTorch is tracking the operations performed to compute out for potential gradient calculations later.

tensor([[-0.1262,  0.1080, -0.1792]], grad_fn=<AddmmBackward0>)


The output `tensor([[-0.1262, 0.1080, -0.1792]], grad_fn=<AddmmBackward0>)` represents the **logits** produced by the neural network for the single input sample.

Here's what each part means:

`tensor([[-0.1262, 0.1080, -0.1792]])`: This is the actual output data. It's a tensor with **one row and three columns**. Since the NeuralNetwork was instantiated with num_outputs=3, each of these three values **corresponds to the raw, unnormalized score (or logit) **for each of the three possible output classes. A higher logit value for a class generally indicates that the model has higher confidence that the input belongs to that class.
`grad_fn=<AddmmBackward0`>: This part indicates that this tensor is the result of an operation (`AddmmBackward0` corresponds to a matrix multiplication followed by an addition, which is what happens in a linear layer), and that PyTorch is tracking the operations performed to create this tensor. This tracking is essential for automatic differentiation, allowing PyTorch to compute gradients during the backward pass for training.

In a typical classification task, these logits would be passed through a softmax function to convert them into probabilities for each class, which sum up to 1.

结果中返回的 3 个数值对应于分配给每个输出节点的分数。注意输出张量还包含了一个grad_fn 值.

`grad_fn=<AddmmBackward0>`意味着我们正在查看的张量是通过矩阵乘法和加法操作创建的。PyTorch 会在反向传播期间使用这些信息来计算梯度。`grad_fn=<AddmmBackward0>`中的
`<AddmmBackward0>`指定了执行的操作。在这种情况下，它执行的是一个 Addmm 操作。Addmm代表的是矩阵乘法（mm）后接加法（Add）的组合运算

In [ ]:
'''
如果只想使用网络进行预测而不进行训练或反向传播（比如在训练之后使用它进行预测），
那么为反向传播构建这个计算图可能会浪费资源，因为它会执行不必要的计算并消耗额外的内
存。因此，当使用模型进行推理（比如做出预测）而不是训练时，最好的做法是使用
torch.no_grad()上下文管理器。这会告诉 PyTorch 无须跟踪梯度，从而可以显著节省内存和
计算资源
'''
with torch.no_grad():
    out = model(X)
print(out)

tensor([[-0.1262,  0.1080, -0.1792]])


In [ ]:
with torch.no_grad():
    out = torch.softmax(model(X), dim=1) #dim=1: This specifies the dimension along which the softmax function is applied. In this case, dim=1 means the softmax is applied across the columns (the different output classes) for each input sample (row). This is the standard way to apply softmax for classification tasks where each row represents a single instance and each column represents a class score.
print(out) # 输出的结果现在这些值可以解释为类别成员的概率，并且它们的总和大约为 1

tensor([[0.3113, 0.3934, 0.2952]])


**模型通常返回 Logits:** 在 PyTorch 中构建分类模型时，习惯上让模型的最后一层只进行线性变换，输出的是原始的、未经过 Softmax 或 Sigmoid 激活的数值，这些数值被称为 logits。
损失函数中集成了 Softmax/Sigmoid: PyTorch 提供了一些常用的损失函数，例如用于多分类的 CrossEntropyLoss 和用于二分类的 BCEWithLogitsLoss。这些损失函数在内部已经集成了 Softmax（或 Sigmoid）操作和负对数似然损失计算。

**原因：** 数值计算的效率和稳定性: 将 Softmax/Sigmoid 和损失函数结合在一起计算，相比于先单独计算 Softmax/Sigmoid 再计算损失，在数值计算上更有效率，并且可以提高数值稳定性。这是因为 Softmax/Sigmoid 的计算涉及到指数运算，当输入值很大或很小时，可能会出现数值溢出或下溢的问题。将它们与对数似然损失结合后，可以通过一些数学技巧（例如 LogSumExp 技巧）来避免这些问题，从而提高计算的精度和稳定性。

**需要概率时再显式调用:** 因此，如果您只是想在训练过程中计算损失并进行反向传播，您可以直接将模型的 logits 输出传递给 PyTorch 集成了 Softmax/Sigmoid 的损失函数。只有当您需要查看模型对每个类别的预测概率时（例如在推理阶段或者进行模型评估时），才需要显式地对模型的 logits 输出应用 Softmax 函数，就像您在代码中看到的那样。

总结来说，这种做法是为了在训练过程中获得更好的数值表现，同时仍然允许您在需要时方便地获取类别概率。

## A.6 设置高效的数据加载器

<img src="https://raw.githubusercontent.com/MLNLP-World/LLMs-from-scratch-CN/main/imgs/A6/1.png" width="600px">

In [ ]:
#首先创建一个简单的示例数据集，其中包含 5 个训练示例，每个示例有两个特征
X_train = torch.tensor([
    [-1.2, 3.1],
    [-0.9, 2.9],
    [-0.5, 2.6],
    [2.3, -1.1],
    [2.7, -1.5]
])
'''
PyTorch 要求类别标签从标签 0 开始，并且最大的类别标签值不得超过输出节点数减 1
（因为 Python 的索引从 0 开始）。因此，如果我们有类别标签 0、1、2、3 和 4，那么神经
网络的输出层应包含 5 个节点。
'''
#与训练示例一起，我们还创建了一个包含相应类别标签的张量：3 个示例属于类别标签 0，两个示例属于类别标签 1
y_train = torch.tensor([0, 0, 0, 1, 1])

In [ ]:
#还构建了一个包含两个样本的测试集
X_test = torch.tensor([
    [-0.8, 2.8],
    [2.6, -1.6],
])

y_test = torch.tensor([0, 1])

In [ ]:
from torch.utils.data import Dataset #通过继承 PyTorch 的 Dataset 父类来创建一个自定义数据集类 ToyDataset


class ToyDataset(Dataset):
    def __init__(self, X, y):
        self.features = X
        self.labels = y

    def __getitem__(self, index): #检索一条数据记录及其对应标签的说明
        one_x = self.features[index]
        one_y = self.labels[index]
        return one_x, one_y

    def __len__(self): #检索一条数据记录及其对应标签的说明
        return self.labels.shape[0]

train_ds = ToyDataset(X_train, y_train)
test_ds = ToyDataset(X_test, y_test)

上述代码片段定义了一个自定义的数据集类 `ToyDataset`，它继承自 PyTorch 的 `torch.utils.data.Dataset`。通过继承 `Dataset`，您的类可以使用 PyTorch 内置的数据处理功能，并能与 `DataLoader` 很好地集成。

以下是该类及其方法的详细解释：

1.  **`class ToyDataset(Dataset):`**: 这行代码定义了一个名为 `ToyDataset` 的新类，它继承自 `torch.utils.data.Dataset`。继承 `Dataset` 类可以让您的自定义数据集具备 PyTorch 数据集的基本功能。
2.  **`def __init__(self, X, y):`**: 这是 `ToyDataset` 类的构造函数。它接收两个参数：
    *   `X`: 预计是包含您的特征（输入数据）的张量或其他数据结构。
    *   `y`: 预计是包含您的标签（目标输出）的张量或其他数据结构。
    *   在构造函数内部，这些输入的 `X` 和 `y` 被存储为实例变量 `self.features` 和 `self.labels`。
3.  **`def __getitem__(self, index):`**: 这是一个特殊的方法，它允许您使用索引（例如 `dataset[0]`）来访问数据集中的单个样本。它接收一个 `index` 作为输入。
    *   `one_x = self.features[index]`: 这行代码检索给定 `index` 处样本的特征数据。
    *   `one_y = self.labels[index]`: 这行代码检索给定 `index` 处样本的标签数据。
    *   `return one_x, one_y`: 该方法返回一个元组，其中包含请求索引处的特征和标签。`DataLoader` 就是通过这个方法从您的数据集中检索单个样本的。
4.  **`def __len__(self):`**: 这是另一个特殊的方法，它允许您使用 `len()` 函数（例如 `len(dataset)`）来获取数据集中的样本总数。
    *   `return self.labels.shape[0]`: 这行代码返回标签张量（`self.labels.shape[0]`）的第一个维度的大小，这对应于数据集中的样本总数。

在定义了类之后，接下来的两行代码：
*   **`train_ds = ToyDataset(X_train, y_train)`**: 使用训练数据 `X_train` 和 `y_train` 创建了一个 `ToyDataset` 的实例，并将其赋值给 `train_ds` 变量。
*   **`test_ds = ToyDataset(X_test, y_test)`**: 使用测试数据 `X_test` 和 `y_test` 创建了一个 `ToyDataset` 的实例，并将其赋值给 `test_ds` 变量。

现在，`train_ds` 和 `test_ds` 对象就是 PyTorch 的 `Dataset` 对象，可以与 `DataLoader` 一起使用，以便在训练和评估期间高效地批量加载数据。

In [ ]:
len(train_ds)

5

In [ ]:
from torch.utils.data import DataLoader

torch.manual_seed(123)

train_loader = DataLoader(
    dataset=train_ds,
    batch_size=2,
    shuffle=True,
    num_workers=0
)

上述代码片段使用 `torch.utils.data.DataLoader` 为之前创建的 `train_ds` 数据集构建了一个数据加载器。数据加载器是 PyTorch 中用于高效加载和批量处理数据的重要工具。

以下是代码的详细解释：

1.  **`from torch.utils.data import DataLoader`**: 导入 PyTorch 中用于创建数据加载器的 `DataLoader` 类。
2.  **`torch.manual_seed(123)`**: 设置 PyTorch 随机数生成器的种子。这确保了数据加载器在进行随机操作（例如打乱数据）时，结果是可重复的。
3.  **`train_loader = DataLoader(...)`**: 创建一个 `DataLoader` 实例，并将其赋值给 `train_loader` 变量。
    *   **`dataset=train_ds`**: 指定要加载数据的数据集对象，这里使用的是之前创建的 `train_ds`。
    *   **`batch_size=2`**: 设置每个批次的数据大小为 2。这意味着数据加载器每次迭代将返回 2 个训练样本及其对应的标签。
    *   **`shuffle=True`**: 设置为 `True` 表示在每个 epoch（训练周期）开始时，数据加载器会打乱数据集中的数据顺序。这有助于防止模型学习到数据中的顺序模式，提高模型的泛化能力。
    *   **`num_workers=0`**: 设置用于数据加载的子进程数。`0` 表示数据加载将在主进程中进行。对于小型数据集或在调试时，设置为 `0` 通常足够。对于大型数据集，可以增加 `num_workers` 以利用多核处理器加速数据加载。

创建 `train_loader` 后，您就可以在训练循环中迭代它，每次迭代都会为您提供一个包含指定批次大小的数据和标签的张量。

In [ ]:
test_ds = ToyDataset(X_test, y_test)

test_loader = DataLoader(
    dataset=test_ds,
    batch_size=2,
    shuffle=False,
    num_workers=0
)

根据下面的输出，可以看到 train_loader 迭代了训练数据集，每个训练示例正好访问一次。
这被称为一个训练轮次。由于我们使用 torch.manual_seed(123)设置了随机数生成器，因此
你应该得到完全相同的训练示例打乱顺序。然而，当你再次迭代数据集时，你会发现打乱的顺序
已经发生变化。这是为了防止深度神经网络在训练过程中陷入重复更新循环

In [ ]:
#在实例化训练数据加载器后，可以对其进行迭代。对 test_loader 的迭代与之类似，但为简洁起见，这里省略了具体细节：
for idx, (x, y) in enumerate(train_loader):
    print(f"Batch {idx+1}:", x, y)
#在这里指定的批次大小为 2，但第三批次仅包含一个示例。这是因为我们有 5 个训练示例，而 5 不能被 2 整除

Batch 1: tensor([[ 2.3000, -1.1000],
        [-0.9000,  2.9000]]) tensor([1, 0])
Batch 2: tensor([[-1.2000,  3.1000],
        [-0.5000,  2.6000]]) tensor([0, 0])
Batch 3: tensor([[ 2.7000, -1.5000]]) tensor([1])


In [ ]:
#在实践中，如果一个训练轮次的最后一个批次显著小于其他批次，那么可能会影响训练过程中的收敛。为此，可以设置 drop_last=True，这将在每轮中丢弃最后一个批次

train_loader = DataLoader(
    dataset=train_ds,
    batch_size=2,
    shuffle=True,
    num_workers=0,
    drop_last=True #drop_last=True，这将在每轮中丢弃最后一个批次
)

In [ ]:
for idx, (x, y) in enumerate(train_loader):
    print(f"Batch {idx+1}:", x, y) #可以看到最后一个批次被省略了

Batch 1: tensor([[-1.2000,  3.1000],
        [-0.5000,  2.6000]]) tensor([0, 0])
Batch 2: tensor([[ 2.3000, -1.1000],
        [-0.9000,  2.9000]]) tensor([1, 0])


**DataLoader 中的 num_workers=0 设置**:

这个参数在 PyTorch 的DataLoader 函数中对于并行加载和预处理数据至关重要。当 num_workers 设置为 0 时，数据加载将在主进程而不是单独的工作进程中进行。这看起来似乎没有问题，但在使用** GPU 训练较大的网络**时，这可能会**导致模型训练显著减慢**。这是因为 CPU **不仅要处理深度学习模型**，还要**花时间加载和预处理数据**。因此，GPU 在等待 CPU 完成这些任务时可能会闲置。相反，当**num_workers 设置为大于 0** 的数值时，会**启动多个工作进程**并行加载数据，从而**释放主进程专注于训练模型**


<img src="https://raw.githubusercontent.com/MLNLP-World/LLMs-from-scratch-CN/main/imgs/A6/2.png" width="600px">

在没有多个工作进程的情况下加载数据（设置 num_workers=0）会导致数据加载瓶
颈，模型会在下一个批次加载完成前处于空闲状态（左）。如果启用多个工作进程，那
么数据加载器可以在后台排队下一个批次的数据（右）

## A.7 典型的训练循环

In [ ]:
import torch.nn.functional as F


torch.manual_seed(123)
model = NeuralNetwork(num_inputs=2, num_outputs=2) # 该数据集有两个特征和两个类别
optimizer = torch.optim.SGD(model.parameters(), lr=0.5) # 优化器需要知道哪些参数需要优化

num_epochs = 3

for epoch in range(num_epochs):

    model.train()
    for batch_idx, (features, labels) in enumerate(train_loader):

        logits = model(features)

        loss = F.cross_entropy(logits, labels) # 损失函数

        optimizer.zero_grad() #将上一轮的梯度置 0，以防止意外的梯度累积
        loss.backward() #根据模型参数计算损失的梯度

        optimizer.step() #优化器使用梯度更新模型参数

        ### 日志
        print(f"Epoch: {epoch+1:03d}/{num_epochs:03d}"
              f" | Batch {batch_idx:03d}/{len(train_loader):03d}"
              f" | Train/Val Loss: {loss:.2f}")

    model.eval()
    # 插入可选的模型评估代码

Epoch: 001/003 | Batch 000/002 | Train/Val Loss: 0.75
Epoch: 001/003 | Batch 001/002 | Train/Val Loss: 0.65
Epoch: 002/003 | Batch 000/002 | Train/Val Loss: 0.44
Epoch: 002/003 | Batch 001/002 | Train/Val Loss: 0.13
Epoch: 003/003 | Batch 000/002 | Train/Val Loss: 0.03
Epoch: 003/003 | Batch 001/002 | Train/Val Loss: 0.00


上述代码片段实现了一个典型的 PyTorch 模型训练循环。它涵盖了神经网络训练的核心步骤，包括前向传播、计算损失、反向传播和参数更新。

以下是代码的详细解释：

1.  **`import torch.nn.functional as F`**: 导入 PyTorch 中的函数式接口，通常包含损失函数、激活函数等。这里主要用于使用 `F.cross_entropy`。
2.  **`torch.manual_seed(123)`**: 设置 PyTorch 随机数生成器的种子，确保模型初始化和数据加载（如果 `DataLoader` 使用随机性）的可重复性。
3.  **`model = NeuralNetwork(num_inputs=2, num_outputs=2)`**: 实例化之前定义的 `NeuralNetwork` 模型。这里输入特征数为 2，输出类别数为 2。
4.  **`optimizer = torch.optim.SGD(model.parameters(), lr=0.5)`**: 创建一个随机梯度下降（SGD）优化器。
    *   `model.parameters()`: 这会获取模型中所有需要梯度的参数（即需要训练的权重和偏置）。
    *   `lr=0.5`: 设置学习率（learning rate）为 0.5，控制每次参数更新的步长。
5.  **`num_epochs = 3`**: 设置训练的总轮次（epochs）为 3。一个 epoch 表示模型遍历整个训练数据集一次。
6.  **`for epoch in range(num_epochs):`**: 外层循环，控制训练的 epoch 数量。
7.  **`model.train()`**: 将模型设置为训练模式。这会启用 dropout 和 batch normalization 等在训练和评估阶段行为不同的层。
8.  **`for batch_idx, (features, labels) in enumerate(train_loader):`**: 内层循环，迭代 `train_loader` 获取每个批次的数据和标签。
    *   `enumerate(train_loader)`: 同时获取批次的索引 (`batch_idx`) 和数据 (`features`, `labels`)。
    *   `(features, labels)`: `DataLoader` 每次迭代返回一个批次的特征和对应的标签。
9.  **`logits = model(features)`**: **前向传播**。将当前批次的特征 `features` 输入到模型中，获取模型的原始输出（logits）。
10. **`loss = F.cross_entropy(logits, labels)`**: **计算损失**。使用交叉熵损失函数计算模型输出 `logits` 与真实标签 `labels` 之间的损失。`F.cross_entropy` 内部已经集成了 Softmax 操作。
11. **`optimizer.zero_grad()`**: **梯度清零**。在计算当前批次的梯度之前，将之前计算的梯度清零。这是必要的，因为 PyTorch 默认会将梯度累加。
12. **`loss.backward()`**: **反向传播**。计算损失关于模型参数的梯度。
13. **`optimizer.step()`**: **参数更新**。根据计算出的梯度和优化器的设置（例如学习率），更新模型的参数。
14. **`### 日志`**: 代码中的注释，表示接下来的代码用于打印训练过程中的信息。
15. **`print(f"Epoch: {epoch+1:03d}/{num_epochs:03d}" ...)`**: 打印当前 epoch、批次信息以及训练损失。
16. **`model.eval()`**: 将模型设置为评估模式。这会禁用 dropout 并将 batch normalization 层设置为使用整个数据集的均值和方差。在计算验证集或测试集的指标之前，务必设置为评估模式。
17. **`# 插入可选的模型评估代码`**: 注释，提示这里可以插入计算模型在验证集上的性能（例如准确率）的代码。

这个训练循环是大多数 PyTorch 项目的基础框架。通过重复执行前向传播、损失计算、反向传播和参数更新，模型逐渐学习从输入到输出的映射。

**练习3:** 上述代码中介绍的神经网络有多少个参数？

In [ ]:
# here use model = NeuralNetwork(num_inputs=2, num_outputs=2) # 该数据集有两个特征和两个类别
# then back to previous NeuralNetwork
model = NeuralNetwork(num_inputs=2, num_outputs=2)
num_params = sum(p.numel() for p in model.parameters() if p.requires_grad)
print("Total number of trainable model parameters:", num_params)

在实际操作中，通常会使用第三个数据集，即所谓“验证数据集”，来找到最优的超参数设置。验证集类似于测试集。然而，虽然只想精确地使用一次测试集以避免评估偏差，但通常会多次使用验证集来调整模型设置。

In [ ]:
# 在训练好模型后，可以使用它进行预测
model.eval()

with torch.no_grad():
    outputs = model(X_train)

print(outputs)

tensor([[ 2.8569, -4.1618],
        [ 2.5382, -3.7548],
        [ 2.0944, -3.1820],
        [-1.4814,  1.4816],
        [-1.7176,  1.7342]])


In [ ]:
# 为了获得类别成员概率，可以使用 PyTorch 的 softmax 函数
torch.set_printoptions(sci_mode=False)
probas = torch.softmax(outputs, dim=1)
print(probas)  #第一个值（列）表示该训练示例属于类别标签 0 的概率为 99.91%，属于类别标签 1 的概率为 0.09%。（这里使用 set_printoptions 是为了让输出更加易读。）

predictions = torch.argmax(probas, dim=1) #This function returns the index of the maximum value along a specified dimension.
#argmax 函数将这些概率值转换为类别标签预测。如果设置 dim=1，它将返回每行中最大值的索引位置（设置 dim=0 则返回每列中最大值的索引位置）
print(predictions)

tensor([[    0.9991,     0.0009],
        [    0.9982,     0.0018],
        [    0.9949,     0.0051],
        [    0.0491,     0.9509],
        [    0.0307,     0.9693]])
tensor([0, 0, 0, 1, 1])


In [ ]:
# 请注意，为了获得类别标签，计算 softmax 概率并非必需步骤，也可以直接对 logits（输出）应用argmax 函数
'''
为什么可以直接对 logits 使用 argmax？
因为 softmax 函数是一个单调递增函数。这意味着如果一个类别在原始 logits 中的得分最高，
那么它在经过 softmax 转换后的概率也一定是最高的。
因此，为了找到概率最高的类别，我们只需要找到 logits 中得分最高的类别对应的索引即可，
无需额外的 softmax 计算。这在计算上更有效率。
'''
predictions = torch.argmax(outputs, dim=1)
print(predictions)

tensor([0, 0, 0, 1, 1])


In [ ]:
predictions == y_train

tensor([True, True, True, True, True])

In [ ]:
torch.sum(predictions == y_train) #使用 torch.sum 可以计算正确预测的数量

tensor(5)

In [ ]:
def compute_accuracy(model, dataloader):
  '''
  用于计算神经网络模型在给定数据集上的准确率
  model: 要评估的神经网络模型。
  dataloader: 用于提供数据的 DataLoader 对象（例如测试集或验证集的 DataLoader）。
  '''
    # 这行将模型设置为评估模式。在评估模式下，像 dropout 和 batch normalization 这样的层会表现得与训练模式不同（例如，dropout 会被禁用，batch normalization 会使用整个数据集的统计信息）。在计算准确率或进行推理时，务必将模型设置为评估模式
    model = model.eval()
    correct = 0.0
    total_examples = 0

    for idx, (features, labels) in enumerate(dataloader):

        with torch.no_grad(): #这是一个上下文管理器，用于指示 PyTorch 在此代码块内部的操作中不要跟踪梯度。在评估或推理阶段，我们不需要计算梯度，使用 torch.no_grad() 可以节省内存和计算资源。
            logits = model(features) #将当前批次的特征 features 输入到模型中进行前向传播，得到模型的原始输出（logits）

        predictions = torch.argmax(logits, dim=1)
        compare = labels == predictions
        correct += torch.sum(compare)
        total_examples += len(compare)

    return (correct / total_examples).item() #正确预测的比例是一个介于 0 和 1 之间的值。调用.item()会将张量的值以 Python 浮点数的形式返回


In [ ]:
compute_accuracy(model, train_loader)

1.0

In [ ]:
compute_accuracy(model, test_loader)

1.0

## A.8 保存和加载模型

In [ ]:
# 下面是在PyTorch 中保存和加载模型的推荐方法
torch.save(model.state_dict(), "model.pth")
# 模型的 state_dict 是一个 Python 字典对象，它可以将模型中的每一层映射到其可训练参数（权重和偏置）
# model.pth 是保存到磁盘的模型文件的任意文件名，我们可以使用任何名称和文件后缀，不过.pth 和.pt 是最常见的约定

In [ ]:
# 保存模型后，可以从磁盘中恢复它
model = NeuralNetwork(2, 2) # 需要与最初保存的模型完全匹配
model.load_state_dict(torch.load("model.pth", weights_only=True))
# torch.load("model.pth")函数读取文件 model.pth，并重建包含模型参数的 Python 字典对象
# model.load_state_dict()则将这些参数应用到模型中，有效地恢复了我们保存模型时模型的学习状态

<All keys matched successfully>

**为什么在同一会话中 `model = NeuralNetwork(2, 2)` 不是严格必需的？**

在您执行以下代码时：

```python
torch.save(model.state_dict(), "model.pth")
```

`model` 变量已经指向了您之前创建和训练好的那个 `NeuralNetwork(2, 2)` 模型的实例。

如果您在**同一个 Python 会话中**紧接着执行：

```python
model.load_state_dict(torch.load("model.pth", weights_only=True))
```

并且在保存模型后**没有对 `model` 变量进行重新赋值或改变其指向的对象**，那么 `model` 仍然是那个已经存在的模型实例。

在这种情况下，**直接调用 `model.load_state_dict()` 就可以将从文件中加载的参数应用到这个现有的模型实例上**。因此，**再次创建 `model = NeuralNetwork(2, 2)` 并非强制性的。**

---

**为什么注释中包含了 `model = NeuralNetwork(2, 2)`，并且强调架构必须匹配？**

为了说明需要模型实例：

注释解释了为什么要包含：

```python
model = NeuralNetwork(2, 2)
```

即使在同一会话中不是必需的，它仍然是为了演示一个**通用的、跨会话的加载模型的流程**。

当您在一个**新的 Python 会话或程序中加载模型**时，您不可能直接调用 `model.load_state_dict()`，因为此时内存中还没有一个名为 `model` 的神经网络实例。

您必须先**创建一个与您保存时完全相同架构的模型实例**，然后才能将保存的参数加载到这个新的实例中。

---

强调架构匹配：

`model.load_state_dict()` 方法是将一个字典（包含参数名称和对应的张量值）加载到一个模型实例中。

这个过程是**基于参数名称和形状进行的**。如果新创建的模型实例的架构与保存时的模型不匹配（例如，层数不同、每层的节点数不同），那么参数的名称或形状就会不匹配，`load_state_dict` 方法就会报错，无法成功加载参数。

因此，**确保加载模型的架构与保存模型的架构完全一致是至关重要的。**

---

总结：

这段注释的目的是为了向您展示一个**通用的模型加载过程**，即使在某些特定情况下（如在同一会话中紧接着加载），**创建新的模型实例那一步可以省略**。

但更重要的是强调：

> **无论何时加载模型，都必须先在内存中创建一个与保存时架构完全一致的模型实例，然后才能使用 `load_state_dict` 方法将保存的参数加载到这个实例中。**

## A.9 使用 GPU 优化训练性能

### A.9.1 在GPU设备运行PyTorch

见 [code-part2.ipynb](code-part2.ipynb)

### A.9.2 单个GPU训练

见 [code-part2.ipynb](code-part2.ipynb)

### A.9.3 使用多个GPU训练

见 [DDP-script.py](DDP-script.py)